storage/src/key_value/agent/block.rs

```diff
@@ -0,0 +1,364 @@
+        match location {
+            Some(location) => Ok(Some(TransactionLocation::read(&location[..])?)),
+            None => Ok(None),
+        }
+    }
+
+    pub(super) fn get_canon_blocks(&mut self, limit: Option<u32>) -> Result<Vec<SerialBlock>> {
+        let index = self.inner.get_column(KeyValueColumn::BlockIndex)?
+            .into_iter()
+            .filter(|(key, _)| key.len() == 4) // only interested in block index -> block hash maps
+            .map(|(key, value)| (Self::read_u32(&key[..]).expect("invalid key"), value[..].into()))
+            .collect::<Vec<(u32, Digest)>>();
+        let mut blocks: HashMap<u32, SerialBlock> = HashMap::new();
+        let mut max_block_number = 0u32;
+        for (block_number, hash) in index {
+            if let Some(limit) = limit {
+                if block_number > limit {
+                    continue;
+                }
+            }
+            if block_number > max_block_number {
+                max_block_number = block_number;
+            }
+            let block = self.get_block(&hash)?;
+            blocks.insert(block_number, block);
+        }
+        let mut out = Vec::with_capacity(max_block_number as usize + 1);
+        for i in 0..=max_block_number {
+            out.push(
+                blocks
+                    .remove(&i)
+                    .ok_or_else(|| anyhow!("missing block {} for get_canon_blocks", i))?,
+            );
+        }
+        Ok(out)
+    }
+
+    pub(super) fn get_block_hashes(&mut self, limit: Option<u32>, filter: BlockFilter) -> Result<Vec<Digest>> {
+        let mut hashes = match filter {
+            BlockFilter::CanonOnly => {
+                let index = self.inner.get_column_keys(KeyValueColumn::BlockIndex)?
+                    .into_iter()
+                    .filter(|key| key.len() != 4) // only interested in block hash keys
+                    .map(|key| key[..].into())
+                    .collect::<Vec<Digest>>();
+                index
+            }
+            BlockFilter::NonCanonOnly => {
+                let all = self.get_block_hashes(None, BlockFilter::All)?;
+                let canon = self
+                    .get_block_hashes(None, BlockFilter::CanonOnly)?
+                    .into_iter()
+                    .collect::<HashSet<Digest>>();
+                let non_canon = all.into_iter().filter(|hash| !canon.contains(hash)).collect();
+                non_canon
+            }
+            BlockFilter::All => {
+                let hashes = self
+                    .inner
+                    .get_column_keys(KeyValueColumn::BlockHeader)?
+                    .into_iter()
+                    .map(|key| key[..].into())
+                    .collect::<Vec<Digest>>();
+                hashes
+            }
+        };
+        // this isn't the most efficient way to do limit here, but it's good enough and clean
+        if let Some(limit) = limit {
+            hashes.truncate(limit as usize);
```

Comment:
this is fine, but are the blocks sorted by their index? otherwise we're likely to get them in random order and potential gaps, which seems less practical